### DISCUSSION TITLE
DataGrid Expand/Contract Button for DetailRowTemplate

### DISCUSSION THREAD

**Discussion by jeffbvox** [Megabit/Blazorise#2467] [Q&A] [answered]
Is there any built in Expand/Contract button for showing and hiding the DetailRowTemplate data instead of clicking on the row and checking the DetailRowTrigger result? If not, are there any workarounds for having one or are there plans to have it as an option in the future? Most data grids have this type of functionality and I think this is one of the only features the Blazorise DataGrid is missing.

Thanks!

Jeff B.

**Comment by stsrki** [comment, marked as answer]
The detail row is activated with `DetailRowTrigger` callback. Meaning you can check for anything inside of it and as long as it returns true it will show the detail row.

```cshtml
DetailRowTrigger="@((item)=>item.Id == selectedEmployee?.Id && somethingElseOnYourPage == true)"
```

**Reply by jeffbvox** [reply]
I was wondering if there was a single UI element that could control the expansion like the circled items:

![image](https://user-images.githubusercontent.com/47092030/121573172-eec3a580-c9f2-11eb-8edb-a1dc5afdbb32.png)

I could probably add a column to have a button that controls the expansion/contraction of the detail row using the DetailRowTrigger callback, but I don't see a way to detect which button/cell was used to trigger the callback. If there was a way to determine which cell triggered it, then I could suppress it for the other non-button cells. 

I was thinking it might be something that is built in that I missed.

**Reply by stsrki** [reply]
There is not built-in expansion component like this. We might implement it in the future but for now, it's not so much of a prioroty.

---

I can think of several ways to do it.

1. Have a DisplayTemplate with a button that can control a field inside of TItem model. `Clicked=@(()=>context.Item.Expanded = !context.Item.Expanded)`
2. Have a list of expanded items ids, and then on the button clicked event add or remove them from the list.

In either case, you just check for true inside of DetailRowTrigger.

**Comment by jeffbvox** [comment]
Thank you for the tips on how to implement the functionality! I will give it a try.

**Reply by djh2004** [reply]
Jeff, what ultimately worked for you? I am having the same issue. I am trying the suggestions above but need more detail.

Thanks,
   D-J
